ament_cpplint/ament_cpplint/main.py

```diff
@@ -199,27 +205,60 @@ def custom_error(filename, linenum, category, confidence, message):
     return 1 if _cpplint_state.error_count else 0
 
 
-def get_file_groups(paths, extensions):
+def is_file_or_directory_to_exclude(path_to_file, excludes):
+    if not os.path.isdir(path_to_file):
+        path, file_name = os.path.split(path_to_file)
+    else:
+        path = path_to_file
+        file_name = ''
+    path_to_check = pathlib.Path(path)
+
+    def check_path_parents(parents, exclude):
+        for p in parents:
+            if p.match(exclude):
+                return True
+        return False
+
+    for excl in excludes:
+        if (
+            excl == file_name or
+            path_to_check.match(excl) or
+            check_path_parents(path_to_check.parents, excl)
+        ):
+            return True
+    return False
+
+
+def get_file_groups(paths, extensions, excludes):
     # dict mapping root path to files
     groups = {}
     for path in paths:
         if os.path.isdir(path):
             for dirpath, dirnames, filenames in os.walk(path):
+                if is_file_or_directory_to_exclude(dirpath, excludes):
+                    continue
+
                 if 'AMENT_IGNORE' in dirnames + filenames:
                     dirnames[:] = []
                     continue
                 # ignore folder starting with . or _
                 dirnames[:] = [d for d in dirnames if d[0] not in ['.', '_']]
-                dirnames.sort()
+                dirnames[:] = [
+                    d for d in dirnames if not is_file_or_directory_to_exclude(
+                        os.path.join(dirpath, d), excludes)
+                ]
 
+                dirnames.sort()
                 # select files by extension
                 for filename in sorted(filenames):
                     _, ext = os.path.splitext(filename)
                     if ext in ('.%s' % e for e in extensions):
-                        append_file_to_group(groups,
-                                             os.path.join(dirpath, filename))
+                        if filename not in excludes:
```

Comment:
This condition still doesn't make sense.

Please write a unit test which tests a variety of cases to validate the implementation against.

Reply:
humm, why not? the excludes argument might contains a filename

Reply:
I don't think an exclude of `bar` should match a file `foo/bar`. Instead the exclude should state `foo/bar` explicitly. Otherwise how do I exclude the file `bar` but not the file `foo/bar`?

Reply:
should I need to include the test?

Reply:
You don't have to include the test in this PR but as mentioned before writing one will help you to check the logic. Especially the part to consider example file system structure and excludes and what the filtered result is expected to be.

Reply:
@dirk-thomas  I created a gist with a test: https://gist.github.com/ahcorde/bcc558aedad8560cfa46c4b5de366f32

This is the last commit https://github.com/ament/ament_lint/pull/238/commits/6db3ae3adaf9cda7a11dfe23fe279f18c702fb7c

Reply:
I am not sure how to use the gist. `get_file_groups()` seems to always return an empty dict?

Just from reading over the logic I would expect that a file `include/test/cache_unittest.cpp` would be incorrectly skipped even though only `test/cache_unittest.cpp` is specified in the excludes.

Also I still don't see why `is_file_or_directory_to_exclude()` should needs any logic like `check_path_parents()`.